reconcile/openshift_resources_base.py

```diff
@@ -944,6 +959,209 @@ def run(
+                ns, ns_type_overrides
+            )
+            cluster_scoped_types = [
+                k for k in managed_resource_types if not oc.is_kind_namespaced(k)
+            ]
+
+            if len(cluster_scoped_types) > 0:
+                # Check that all non namespaced resources are explicitly set in the
+                # ManagedResourceNames attribute.
+                mrn = ob.get_namespace_resource_names(ns, ns_type_overrides)
+                for kind in cluster_scoped_types:
+                    declared_items = mrn.get(kind, [])
+                    desired_items = set(
+                        self.ri.get_desired_by_type(
+                            cluster_name, ns["name"], kind
+                        ).keys()
+                    )
+                    diff = desired_items.difference(declared_items)
+                    if len(diff) > 0:
+                        errors.append(
+                            CheckError(
+                                "Cluster scoped resources not defined in ManagedResourceNames. "
+                                f"cluster: {cluster_name}, namespace: {ns['name']}, "
+                                f"kind:{kind}, names:{diff}",
+                            )
+                        )
+
+        return errors
+
+
+@dataclass
+class CheckClusterScopedResourceDuplicates:
+    oc_map: OC_Map
+    all_namespaces: Optional[list[Mapping]] = None
+    thread_pool_size: int = 10
+
+    def check(self) -> list[Exception]:
+        errors: list[Exception] = []
+        clusters = set(self.oc_map.clusters() + self.oc_map.clusters(privileged=True))
+        cluster_cs_resources = get_cluster_scoped_resources(
+            self.oc_map, clusters, self.all_namespaces
+        )
+        results = threaded.run(
+            self._remap_cluster_result,
+            list(cluster_cs_resources.items()),
+            self.thread_pool_size,
+            return_exceptions=True,
+        )
+        for r in results:
+            if isinstance(r, Exception):
+                errors.append(r)
+        if errors:
+            return errors
+
+        cluster_resources_lookup = dict(results)
+        for cluster in cluster_resources_lookup:
+            for kind in cluster_resources_lookup[cluster]:
+                for name in cluster_resources_lookup[cluster][kind]:
+                    if len(cluster_resources_lookup[cluster][kind][name]) > 1:
+                        refs = cluster_resources_lookup[cluster][kind][name]
+                        errors.append(
+                            CheckError(
+                                f"Cluster resource defined in multiple namespaces. "
+                                f"cluster: {cluster}, namespaces: {refs}, "
+                                f"kind:{kind}, name:{name}"
+                            )
+                        )
+        return errors
+
+    def _remap_cluster_result(self, input: Tuple[str, Mapping]) -> Tuple[str, Mapping]:
```

Comment:
Perhaps instead of remapping, you could do something like:
```
x = set()
duplicates = list[Tuple[str, str]]

for ns, resources in cluster_resources.items():
    if not ns:
        continue
    for kind, names in resources.items():
        k_ref = remap.setdefault(kind, {})
        for name in names:
            if (kind, name) in x:
                   duplicates.append( (kind, name) )
            else:
                   x.add ( (kind, name) )
return duplicates 
```

If I am not mistaken, this could make the check in CheckClusterScopedResourceDuplicates easier, as you'd have to loop over the duplicates only.

Reply:
Good point. I'm working to "find the duplicates" instead of remapping the whole data. I want to report the involved namespaces in the logs so I'll change this snippet a bit.

Reply:
PTAL